### PR TITLE
Fixed map generate bug

### LINKED ISSUE #75256: Map regenerates after save reloading
### Describe the bug

In https://github.com/CleverRaven/Cataclysm-DDA/releases/tag/cdda-experimental-2024-07-26-0329 visited places become regenerated after reloading save. As result game generates new NPCs, map extras (like wasp nests etc.) in already visited places.
Probably #75188 causes this.

### Attach save file

[Test_-trimmed.tar.gz](https://github.com/user-attachments/files/16394899/Test_-trimmed.tar.gz)

### Steps to reproduce

1. Start new game or load old save.
2. Visit new places (refugee center for example).
3. Save and quit.
4. Load save and go back to visited parts of map.
5. Population of NPCs will be increased and new map extras may appear.

### Expected behavior

Don't regenerated saved map.

### Screenshots

_No response_

### Versions and configuration

- OS: Windows
    - OS Version: 6.1 7601
- Game Version: 2ec51c5 [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Portal Storms Ignore NPCs [personal_portal_storms],
    Slowdown Fungal Growth [no_fungal_growth]
]

### Additional context

_No response_

### LINKED ISSUE #75258: Loading Faction base camp spawns new npcs
### Describe the bug

re- visiting a npc camp (the loggers in this example) spawns another copy of them.

### Attach save file

[Muscle Magic-trimmed.tar.gz](https://github.com/user-attachments/files/16395800/Muscle.Magic-trimmed.tar.gz)

### Steps to reproduce

1. visit a npc camp
2. walk a long distance
3. revisit the same camp

### Expected behavior

npcs should be not regenerated again.

### Screenshots

image

### Versions and configuration

- Os : Opensuse tumbleweed Linux
- Game version : experimental build 2024-07-26-1716
- Graphics version: tiles

### Additional context

_No response_

### LINKED ISSUE #75262: False(?) map notes
### Describe the bug

The map keeps generating map notes of things that aren't there.

Or at least, only partially there. There's _something_ casting a shadow on the ground in this spot where the map says there's a wasp's nest:
image

These notes seem to appear over time in areas I've already explored, but I don't know what triggers them.

### Attach save file

[New Mind-trimmed.tar.gz](https://github.com/user-attachments/files/16397323/New.Mind-trimmed.tar.gz)

### Steps to reproduce

Unsure when this happens. Possibly just over game time or when moving back and forth.

### Expected behavior

Map notes refer to real things.

### Screenshots

image
There are not three ponds here. There is not even one pond. Based on how they normally spawn, I don't think it's even supposed to be possible for them to be this close together.

image
This portal note overwrote the derelict shed note for the shed that's actually here. There's no portal, but there IS some mycus nearby:
image
No note here, though.

### Versions and configuration

- OS: Windows
    - OS Version: 10.0.19045.4651 (22H2)
- Game Version: c336181 [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Portal Storms Ignore NPCs [personal_portal_storms],
    Slowdown Fungal Growth [no_fungal_growth],
    Mind Over Matter [mindovermatter],
    Bionic Slots [cbm_slots],
    Stats Through Kills [stats_through_kills]
]

### Additional context

Possibly related to #75256?

### LINKED ISSUE #75265: (experimental) npc related segfault on load, duplicate npcs.
### Describe the bug

First time filing a report, forgive any inconsistencies or lack of clarity.

Two saves are involved here: my backup and current save. Not sure how to attach the current save file, as it is crash-on-load and I can't exaclty follow the instructions to send it if it's acting like that - so you'll be getting the backup.

Thankfully, the backup was made only minutes earlier, so don't worry about reproduction.

Not entirely sure what happened? certain NPC's have been... doubled? And saving and reloading near them causes a segfault on load. Saving and reloading one or two map tiles from the doppleganger npcs seems to be fine, but anywhere close is bad news.

tldr: I really don't know what's going on here, and I can only really describe it as magic at this point.

### Attach save file

[marshfield-trimmed.tar.gz](https://github.com/user-attachments/files/16397818/marshfield-trimmed.tar.gz)

### Steps to reproduce

1. enter kayak, and paddle north along the river to either Rubik or mr. Lapin (both are doubled, feel free to check both if you want).
2. exit kayak, and gawk at the pair of Rubiks staring you down.
3. why are there two Rubiks, and why are they outside the castle?
4. you can talk to them, they *seem* functional.
5. that's weird.
6. save and exit while gawking at them.
7. load and crash. (cursed by the dopplegangers!)

### Expected behavior

not to be cursed by the dopplegangers.

### Screenshots

these dorks
Screenshot 2024-07-26 211005

### Versions and configuration

- OS: Windows 11
- Game Version: Cataclysm DDA version 26da18e  (should be last week's experimental)
- Graphics version: Tiles, MSXotto+
- ingame language: English
- Mods Loaded: None (other than the vanilla preset cata always runs)

### Additional context

[logs.zip](https://github.com/user-attachments/files/16397888/logs.zip)

### PR DESCRIPTION
**Pull request by PatrikLundell**
#### Summary
None

#### Purpose of change
Fix #75265, fix #75262, fix #75256, fix #75258, i.e. map featured generated again when they should remain unchanged, including generation of duplicate NPCs.

#### Describe the solution
- Restore a lost negation (lost when optimizing the code). This is what caused the re-generation and duplicate characters.
- Tracked down a bizarre failure of .emplace to work properly, where placement at at least one other index cleared one set earlier. This led to a crash on loading of my save after fixing the "real" bug. Replaced emplace with old school indexing (which the style enforcer probably will scream about).
- Swapped a couple of boolean checks around for a negligible performance gain: normal game play should be the optimal path, with editmap stuff being allowed to cost more.

#### Testing
- Before fixing the bugs:
  - Walked to a car and drove to a Mr Lapin (a lot of them have been generated all over the place), saw him in a single instance (I've not visited his place before), verified via the overmap (only a single character listed for that tile).
  - Drove away a bit and noted a levitating wasp nest roof (which is what causes the shadow in one bug report), and saved when 10 or so overmap tiles away from Mr. Lapin.
  - Loaded the save and drove to Mr. Lapin, and found two copies of him.
- After fixing the bugs:
  - Repeating the above, without finding any levitating wasp nest roof (or any nest at all, for that matter). When returning to Mr. Lapin he hadn't been duplicated (again, verified it via the overmap).

#### Additional context
I don't like the 'emplace' failure one bit.
I tracked it down using debug output to debug.log, and that showed that generate emplace at begin+52 was set to true and there was indeed a submap at the corresponding coordinates.
However, 11 iterations later, when the data for grid_pos 21 was emplaced, the data at grid_pos 52 was suddenly false (grid_pos 21 corresponds to pos (1, 0, -10), so it's the start of the next stack).

### CODE REVIEW

**Comment by akrieger** [comment]
`emplace` inserts into a vector, `[] =` overwrites. If the indexing ever went 'backwards' then you'd shift the generated bits over by 1.

**Comment by akrieger** [comment]
image

**Comment by PatrikLundell** [comment]
Oh, that definitely explains it, as the indices jump based on how coordinates are translated into indices. I thought emplace was equivalent to [] but using approved "modern" syntax.

**Comment by akrieger** [comment]
For 1:1 maps it behaves like you expect. For 1:many maps it will append.

**Comment by qtquazar** [comment]
I'm still getting some phantom wasp nests casting shadows from higher levels in game post-fix, although other map notes now seem accurate.

**Comment by PatrikLundell** [comment]
The bug fix does not deal with past damages, so if the nest roofs were generated before this fix was applied they'd still remain, unfortunately. Thus, I hope this is a case of lingering damage rather than freshly produced issues.

**Comment by AlStrangelove** [comment]
Hi, long time player, first time commenter. It might not be a problem with new games, but i've noticed a bunch of Mapgen issues with the random encounters since this patch. Some examples are the wasp nest roof thing, as mentioned above (which i don't recall having encountered previously in this save), as well as several erroneous entries for other things such as supply drops and the like that don't show up where the autonotes put them.

**Comment by PatrikLundell** [comment]
Yes, lingering issues are not addressed. Levitating roofs shouldn't cause any problems beyond being weird. I've never used auto notes, but my guess is that ones representing overwritten features can simply be deleted manually?

Really bad things would be if something overwrote something that's unique, but I don't think that can happen, because I think major features would just be overwritten by the same or similar features (like one shop replacing another), while things like wasp nests and supply drops are minor "decorations" that can be spawned on any field tile, so if the tile is generated again the feature may not show up (and things may show up where there wasn't anything previously).

**Comment by qtquazar** [comment]
> The bug fix does not deal with past damages, so if the nest roofs were generated before this fix was applied they'd still remain, unfortunately. Thus, I hope this is a case of lingering damage rather than freshly produced issues.

If I understand you correctly: no, unfortunately, I just observed the wasp nest issue happening in a fresh game post fix.

**Comment by PatrikLundell** [comment]
Are you sure it actually happened after the fix, rather than you finding it after the fix (because you did only get close enough to actually see it later)?
I just went of a fairly long road driving journey retreading known territory after having enabled auto notes, and failed to get any such notes or see anything that appeared to be newly generated.

**Comment by qtquazar** [comment]
Positive. I play a very difficult start with world wipes each time, and frequently update experimentals, so I cycle through a lot of worlds.

That said, the invisible floating wasp nest is the *only* issue of its kind I've run into, and only once now since the fix. Others have generated properly. I'll capture it if I see it again.

**Comment by PatrikLundell** [comment]
Unfortunately, I don't think capturing it helps. The problem is that all the info we'd have is that the info on that tile has been wiped and replaced, but nothing to indicate why or how.

I'll try to investigate further, but I'm not too optimistic about it at this stage.

**Comment by qtquazar** [comment]
Thank you for the extra effort. I'll report if I see it again. Having been tech support in a past life, I fully appreciate one anecdotal event post-fix is not a good indicator of any kind of recurring problem :)

(However, given I used to be tech support, I am absolutely sure of what I saw, and that it was indeed post-fix, given I know how annoying this kind of report can be.)
